File: packages/lua-runtime/src/lua/chip.lua
---@meta chip

---@class chip
---OPL3 FM synthesis library for authentic retro game music.
---Provides both low-level note control and high-level pattern/song playback
---using the OPL3 chip (AdLib/Sound Blaster compatible).
local chip = {}

-- ========================================================================
-- Table of Contents
-- ========================================================================
-- 1. Asset Management
-- 2. Lifecycle
-- 3. Direct Note Control
-- 4. Volume / Effects
-- 5. Pattern Building
-- 6. File Loading
-- 7. Playback Control
-- 8. Instrument Bank

-- ========================================================================
-- 1. Asset Management
-- ========================================================================

---Register a directory path to scan for chip music files (.wcol, .wsng).
---Must be called before chip.start().
---@param path string Directory path relative to the script
---@usage chip.add_path("assets")
function chip.add_path(path) end

---Register a chip music file by name.
---Must be called before chip.start().
---@param name string Name to reference the file by
---@param filename string Filename relative to a registered path
---@usage chip.load_file("demo", "demo_song.wcol")
function chip.load_file(name, filename) end

---Initialize the chip engine and load all registered assets.
---Must be called after add_path/load_file and before playback.
---@usage chip.start()
function chip.start() end

-- ========================================================================
-- 2. Lifecycle
-- ========================================================================

---Initialize the chip engine without asset loading.
---Use this for direct note control without pre-loaded files.
---@usage chip.init()
function chip.init() end

---Destroy the chip engine and release resources.
---@usage chip.destroy()
function chip.destroy() end

-- ========================================================================
-- 3. Direct Note Control
-- ========================================================================

---Set the instrument for a track.
---@param track integer Track number (0-17)
---@param id integer Instrument ID from the loaded bank (0-based)
---@usage chip.set_instrument(0, 0) -- Set track 0 to instrument 0 (Piano)
function chip.set_instrument(track, id) end

---Start playing a note on a track.
---@param track integer Track number (0-17)
---@param note string|integer Note name ("C-4", "F#3", "Bb2") or MIDI number (0-127)
---@param velocity? integer Velocity (0-127, default 64)
---@usage chip.note_on(0, "C-4", 64)
---@usage chip.note_on(0, 60, 100) -- MIDI note 60 = C-4
function chip.note_on(track, note, velocity) end

---Stop playing a note on a track.
---If no note is specified, silences all notes on the track.
---@param track integer Track number (0-17)
---@param note? string|integer Note name or MIDI number (optional)
---@usage chip.note_off(0, "C-4")
---@usage chip.note_off(0) -- Silence all notes on track 0
function chip.note_off(track, note) end

---Stop all notes on all tracks.
---@usage chip.all_notes_off()
function chip.all_notes_off() end

-- ========================================================================
-- 4. Volume / Effects
-- ========================================================================

---Set master volume.
---@param vol number Volume level (0.0 to 1.0)
---@usage chip.set_volume(0.5)
function chip.set_volume(vol) end

---Set output gain.
---@param gain number Gain multiplier
---@usage chip.set_gain(1.5)
function chip.set_gain(gain) end

-- ========================================================================
-- 5. Pattern Building
-- ========================================================================

---@class ChipPatternBuilder
---Fluent pattern builder for programmatic music creation.
local ChipPatternBuilder = {}

---Set a note at a specific position.
---@param row integer Row number (0-based)
---@param track integer Track number (0-based)
---@param note string|integer Note name ("C-4") or MIDI number
---@param opts? {instrument?: integer, velocity?: integer, effect?: string}
---@return ChipPatternBuilder self For method chaining
---@usage builder:set_note(0, 0, "C-4", {instrument = 0, velocity = 64})
function ChipPatternBuilder:set_note(row, track, note, opts) end

---Set a note-off at a specific position.
---@param row integer Row number (0-based)
---@param track integer Track number (0-based)
---@return ChipPatternBuilder self For method chaining
---@usage builder:set_note_off(4, 0)
function ChipPatternBuilder:set_note_off(row, track) end

---Build the pattern and load it into the player.
---@usage builder:build()
function ChipPatternBuilder:build() end

---Create a new pattern builder.
---@param tracks integer Number of tracks
---@param rows integer Number of rows
---@param bpm? integer Beats per minute (default 120)
---@return ChipPatternBuilder builder
---@usage local b = chip.pattern(4, 16, 120)
---  b:set_note(0, 0, "C-4", {instrument = 0})
---  b:set_note(4, 0, "E-4")
---  b:build()
function chip.pattern(tracks, rows, bpm) end

-- ========================================================================
-- 6. File Loading
-- ========================================================================

---Load a collection file (.wcol).
---Can pass a pre-loaded asset name or raw YAML string.
---@param name_or_yaml string Asset name or YAML content
---@param song_index? integer Song index within collection (0-based)
---@usage chip.load_collection("demo") -- Load pre-registered asset
---@usage chip.load_collection(yaml_string, 0) -- Load from YAML
function chip.load_collection(name_or_yaml, song_index) end

---Load a song file (.wsng).
---Can pass a pre-loaded asset name or raw YAML string.
---@param name_or_yaml string Asset name or YAML content
---@usage chip.load_song_file("my_song")
function chip.load_song_file(name_or_yaml) end

-- ========================================================================
-- 7. Playback Control
-- ========================================================================

---@class ChipPlayOptions
---@field loop? boolean Whether to loop playback (default true)

---Start playback of the loaded pattern/song.
---@param options? ChipPlayOptions
---@usage chip.play() -- Play with loop
---@usage chip.play({loop = false}) -- Play once
function chip.play(options) end

---Pause playback.
---@usage chip.pause()
function chip.pause() end

---Stop playback and reset position.
---@usage chip.stop()
function chip.stop() end

---Seek to a specific row.
---@param row integer Row number (0-based)
---@usage chip.seek_to_row(16)
function chip.seek_to_row(row) end

---Set the tempo.
---@param bpm integer Beats per minute
---@usage chip.set_bpm(140)
function chip.set_bpm(bpm) end

---@class ChipPlaybackState
---@field playing boolean Whether playback is active
---@field currentRow integer Current playback row
---@field bpm integer Current BPM
---@field totalRows integer Total rows in pattern

---Get the current playback state.
---@return ChipPlaybackState
---@usage local state = chip.get_state()
---  print("Row: " .. state.currentRow)
function chip.get_state() end

---Register a callback for row changes during playback.
---@param cb fun(row: integer) Callback function
---@return fun() unsubscribe Function to remove the callback
---@usage local unsub = chip.on_row_change(function(row)
---  print("Now at row " .. row)
---end)
function chip.on_row_change(cb) end

-- ========================================================================
-- 8. Instrument Bank
-- ========================================================================

---Load an instrument bank from a URL.
---@param url string URL to the GENMIDI JSON file
---@usage chip.load_bank("/instruments/legacy/GENMIDI.json")
function chip.load_bank(url) end

return chip

File: packages/lua-runtime/src/chipLuaCode/chip.ts
/**
 * Standalone chip (OPL3 FM synthesis) Lua code - defines the chip module.
 * This code creates the `_chip` table with all chip functions,
 * delegating to `__chip_*` bridge functions registered in setupChipAPI.ts
 * and `__chip_assets_*` bridge functions registered in setupChipAssetAPI.ts.
 */

export const chipLuaCode = `
    -- ========================================================================
    -- Chip Module (OPL3 FM Synthesis)
    -- ========================================================================
    local _chip = {}

    -- ========================================================================
    -- Note name parsing helper
    -- Supports: "C-4", "C#4", "Db4", "F#3", "Bb2", or raw MIDI numbers
    -- ========================================================================
    local _chip_note_names = { C=0, D=2, E=4, F=5, G=7, A=9, B=11 }

    local function _chip_parse_note(note)
      if type(note) == "number" then return note end
      if type(note) ~= "string" then
        error("Invalid note: expected string or number, got " .. type(note))
      end

      -- Try pattern: letter + optional accidental + optional dash + octave
      -- e.g. "C-4", "C#4", "Db4", "F#3", "Bb2"
      local letter, accidental, octave = note:match("^([A-Ga-g])([#b]?)%-?(%d+)$")
      if not letter then
        -- Try as raw number string
        local num = tonumber(note)
        if num then return num end
        error("Invalid note format: " .. note .. " (expected e.g. 'C-4', 'F#3', or MIDI number)")
      end

      local base = _chip_note_names[letter:upper()]
      if not base then
        error("Invalid note letter: " .. letter)
      end

      if accidental == "#" then
        base = base + 1
      elseif accidental == "b" then
        base = base - 1
      end

      return base + (tonumber(octave) + 1) * 12
    end

    -- ========================================================================
    -- Asset management (follows ail_audio pattern)
    -- ========================================================================
    function _chip.add_path(path)
      __chip_assets_addPath(path)
    end

    function _chip.load_file(name, filename)
      __chip_assets_loadFile(name, filename)
    end

    function _chip.start()
      __chip_assets_start()
      __chip_initAndLoadBank()
    end

    -- ========================================================================
    -- Lifecycle
    -- ========================================================================
    function _chip.init()
      -- __chip_initAndLoadBank is async: awaits player import, init, and bank loading.
      -- wasmoon awaits the returned Promise so Lua blocks until everything is ready.
      __chip_initAndLoadBank()
    end

    function _chip.destroy()
      __chip_destroy()
    end

    -- ========================================================================
    -- Direct note control
    -- ========================================================================
    function _chip.set_instrument(track, id)
      __chip_setInstrument(track, id)
    end

    function _chip.note_on(track, note, velocity)
      local midi = _chip_parse_note(note)
      __chip_noteOn(track, midi, velocity)
    end

    function _chip.note_off(track, note)
      if note ~= nil then
        local midi = _chip_parse_note(note)
        __chip_noteOff(track, midi)
      else
        -- If no note specified, silence all notes on the track
        for n = 0, 127 do
          __chip_noteOff(track, n)
        end
      end
    end

    function _chip.all_notes_off()
      __chip_allNotesOff()
    end

    -- ========================================================================
    -- Volume / effects
    -- ========================================================================
    function _chip.set_volume(vol)
      __chip_setVolume(vol)
    end

    function _chip.set_gain(gain)
      __chip_setGain(gain)
    end

    -- ========================================================================
    -- Pattern building (programmatic)
    -- ========================================================================
    function _chip.pattern(tracks, rows, bpm)
      local handle = __chip_buildPattern(tracks, rows, bpm)
      local builder = {}

      function builder:set_note(row, track, note, opts)
        local midi = _chip_parse_note(note)
        local instrument = opts and opts.instrument or nil
        local velocity = opts and opts.velocity or nil
        local effect = opts and opts.effect or nil
        __chip_patternSetNote(handle, row, track, midi, instrument, velocity, effect)
        return self
      end

      function builder:set_note_off(row, track)
        __chip_patternSetNoteOff(handle, row, track)
        return self
      end

      function builder:build()
        __chip_patternBuild(handle)
      end

      return builder
    end

    -- ========================================================================
    -- File loading (from pre-loaded assets or raw YAML strings)
    -- ========================================================================
    function _chip.load_collection(name_or_yaml, song_index)
      -- Check if it's a pre-loaded asset name
      local content = __chip_assets_getFileContent(name_or_yaml)
      if content then
        __chip_loadCollection(content, song_index)
      else
        -- Treat as raw YAML string
        __chip_loadCollection(name_or_yaml, song_index)
      end
    end

    function _chip.load_song_file(name_or_yaml)
      -- Check if it's a pre-loaded asset name
      local content = __chip_assets_getFileContent(name_or_yaml)
      if content then
        __chip_loadSongFile(content)
      else
        -- Treat as raw YAML string
        __chip_loadSongFile(name_or_yaml)
      end
    end

    -- ========================================================================
    -- Playback control
    -- ========================================================================
    function _chip.play(options)
      local loop = true
      if options then
        if options.loop ~= nil then loop = options.loop end
      end
      __chip_play(loop)
    end

    function _chip.pause()
      __chip_pause()
    end

    function _chip.stop()
      __chip_stop()
    end

    function _chip.seek_to_row(row)
      __chip_seekToRow(row)
    end

    function _chip.set_bpm(bpm)
      __chip_setBPM(bpm)
    end

    function _chip.get_state()
      return __chip_getState()
    end

    function _chip.on_row_change(cb)
      return __chip_onRowChange(cb)
    end

    -- ========================================================================
    -- Instrument bank
    -- ========================================================================
    function _chip.load_bank(url)
      return __chip_loadBankFromUrl(url)
    end

    -- Register chip as a module so require('chip') works
    package.preload['chip'] = function()
      return _chip
    end
`
